File: backend/src/main/java/com/edumetric/backend/enrollment/domain/Enrollment.java
package com.edumetric.backend.enrollment.domain;

import com.edumetric.backend.groups.domain.Group;
import com.edumetric.backend.students.domain.Student;
import com.edumetric.backend.users.domain.User;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;
import java.time.Instant;
import java.time.LocalDate;
import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

@Entity
@Table(name = "enrollments")
@Getter
@Setter
@Builder
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@AllArgsConstructor(access = AccessLevel.PRIVATE)
@EqualsAndHashCode(onlyExplicitlyIncluded = true)
public class Enrollment {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @EqualsAndHashCode.Include
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "student_id", nullable = false)
    private Student student;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "group_id", nullable = false)
    private Group group;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 16)
    private EnrollmentStatus status;

    @Column(name = "enrolled_at", nullable = false)
    private LocalDate enrolledAt;

    @Column(name = "ended_at")
    private LocalDate endedAt;

    @Column(columnDefinition = "TEXT")
    private String reason;

    @ManyToOne(fetch = FetchType.LAZY, optional = true)
    @JoinColumn(name = "created_by_user_id")
    private User createdBy;

    @Column(name = "created_at", nullable = false)
    private Instant createdAt;
}

File: hackathon-front/src/types/api/enrollment.ts
// Mirrors com.edumetric.backend.enrollment DTOs

export type EnrollmentStatus =
  | "ACTIVE"
  | "WITHDRAWN"
  | "TRANSFERRED"
  | "COMPLETED";

export interface EnrollmentDto {
  id: number;
  studentId: number;
  studentName: string;
  groupId: number;
  groupName: string;
  status: EnrollmentStatus;
  enrolledAt: string; // ISO date
  endedAt: string | null;
  reason: string | null;
}

export interface EnrollRequest {
  studentId: number;
  groupId: number;
  reason?: string;
}

export interface TransferRequest {
  studentId: number;
  groupId: number;
  reason?: string;
}

export interface WithdrawRequest {
  studentId: number;
  reason?: string;
}
